pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/ForwardIndexHandler.java

```diff
@@ -271,118 +304,295 @@ private void rewriteRawSVForwardIndex(String column, ColumnMetadata existingColM
+    metadataProperties.put(getKeyFor(column, HAS_DICTIONARY), String.valueOf(true));
+    metadataProperties.put(getKeyFor(column, DICTIONARY_ELEMENT_SIZE),
+        String.valueOf(dictionaryCreator.getNumBytesPerEntry()));
+    updateMetadataProperties(indexDir, metadataProperties);
+
+    // We remove indexes that have to be rewritten when a dictEnabled is toggled. Note that the respective index
+    // handler will take care of recreating the index.
+    removeDictRelatedIndexes(column, segmentWriter);
+
+    // Delete the marker file.
+    FileUtils.deleteQuietly(inProgress);
+
+    LOGGER.info("Created dictionary based forward index for segment: {}, column: {}", segmentName, column);
+  }
+
+  private SegmentDictionaryCreator buildDictionary(String column, ColumnMetadata existingColMetadata,
+      SegmentDirectory.Writer segmentWriter)
+      throws Exception {
+    int numDocs = existingColMetadata.getTotalDocs();
+    // SegmentPartitionConfig is not relevant for rewrites.
+    StatsCollectorConfig statsCollectorConfig =
+        new StatsCollectorConfig(_indexLoadingConfig.getTableConfig(), _schema, null);
+    AbstractColumnStatisticsCollector statsCollector;
+
+    try (ForwardIndexReader reader = LoaderUtils.getForwardIndexReader(segmentWriter, existingColMetadata)) {
+      boolean isSVColumn = reader.isSingleValue();
+
+      switch (reader.getStoredType()) {
+        case INT:
+          statsCollector = new IntColumnPreIndexStatsCollector(column, statsCollectorConfig);
+          break;
+        case LONG:
+          statsCollector = new LongColumnPreIndexStatsCollector(column, statsCollectorConfig);
+          break;
+        case FLOAT:
+          statsCollector = new FloatColumnPreIndexStatsCollector(column, statsCollectorConfig);
+          break;
+        case DOUBLE:
+          statsCollector = new DoubleColumnPreIndexStatsCollector(column, statsCollectorConfig);
+          break;
+        case STRING:
+          statsCollector = new StringColumnPreIndexStatsCollector(column, statsCollectorConfig);
+          break;
+        case BYTES:
+          statsCollector = new BytesColumnPredIndexStatsCollector(column, statsCollectorConfig);
+          break;
+        case BIG_DECIMAL:
+          Preconditions.checkState(isSVColumn, "BigDecimal is not supported for MV columns");
+          statsCollector = new BigDecimalColumnPreIndexStatsCollector(column, statsCollectorConfig);
+          break;
+        default:
+          throw new IllegalStateException();
+      }
+
+      Preconditions.checkState(statsCollector != null);
+      // Note: Special Null handling is not necessary here. This is because, the existing default null value in the
+      // raw forwardIndex will be retained as such while created the dictionary and dict-based forward index. Also,
+      // null value vectors maintain a bitmap of docIds. No handling is necessary there.
+      PinotSegmentColumnReader columnReader =
```

Comment:
We have already initialized the ForwardIndexReader at line 507. We should not need to init another reader here since this will also internally create another forward index reader. 

I suggest we continue to use reader at line 507 to read the values out of forward index and gather stats / build dictionary

Reply:
Just want to clarify that - PinotSegmentColumnReader doesn't initiate a new reader. It just uses the reader that is passed in the constructor. It provides a concise way to read from forward index without worry about dataTypes. Otherwise, we would write redundant code for each dataType.

If you still feel we should not use it, I'm okay with using the existing reader.

Reply:
Not a blocker for me. Ok with existing code as well